core/src/mast/serialization/basic_block_data_decoder.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{
+    AdviceInjector, AssemblyOp, DebugOptions, Decorator, DecoratorList, Operation, SignatureKind,
+};
+
+use super::{decorator::EncodedDecoratorVariant, StringIndex, StringRef};
+use alloc::{string::String, vec::Vec};
+use miden_crypto::Felt;
+use winter_utils::{ByteReader, Deserializable, DeserializationError, SliceReader};
+
+pub struct BasicBlockDataDecoder<'a> {
+    data: &'a [u8],
+    data_reader: SliceReader<'a>,
+    strings: &'a [StringRef],
+}
+
+/// Constructors
+impl<'a> BasicBlockDataDecoder<'a> {
+    pub fn new(data: &'a [u8], strings: &'a [StringRef]) -> Self {
+        let data_reader = SliceReader::new(data);
+
+        Self {
+            data,
+            data_reader,
+            strings,
+        }
+    }
+}
+
+/// Mutators
+impl<'a> BasicBlockDataDecoder<'a> {
+    pub fn decode_operations_and_decorators(
+        &mut self,
+        num_to_decode: u32,
+    ) -> Result<(Vec<Operation>, DecoratorList), DeserializationError> {
```

Comment:
One approach to address the issue you mentioned in https://github.com/0xPolygonMiden/miden-vm/pull/1370#discussion_r1674149012 is to instantiate a separate instance of `SliceReader` for each invocation of this method. The method signature would then take an offset as well and it could look like so:
```rust
pub fn decode_operations_and_decorators(
    &self,
    offset: DataOffset,
    num_to_decode: u32,
) -> Result<(Vec<Operation>, DecoratorList), DeserializationError> {
        let mut operations: Vec<Operation> = Vec::new();
        let mut decorators: DecoratorList = Vec::new();

        let mut data_reader = SliceReader::new(&data[offset as usize..]);
        for _ in 0..num_to_decode {
            let first_byte = data_reader.peek_u8()?;
```
Then, we also can get rid of the `data_reader` field from the `BasicBlockDecoder` struct. It could look something like this:
```rust
pub struct BasicBlockDataDecoder<'a> {
    data: &'a [u8],
    strings: &'a [StringRef],
}
```

Reply:
Done